File: nasdaq_data_link.py
import os
import nasdaqdatalink
import pandas

# Set the API key. The R script uses NASDAQ_API_KEY, so we'll check for that first.
# The nasdaq-data-link package documentation suggests NASDAQ_DATA_LINK_API_KEY.
api_key = os.getenv("NASDAQ_API_KEY")

if api_key:
    nasdaqdatalink.ApiConfig.api_key = api_key
else:
    print("API key not found. Please set the NASDAQ_API_KEY environment variable.")
    exit()

try:
    # Fetch the datatable for ETFG/FUND with the ticker SPY
    # Note: If this returns an empty DataFrame, the 'ETFG/FUND' datatable might not contain GDX data,
    # or GDX might be under a different ticker in this specific datatable.
    # You may need to visit the Nasdaq Data Link website and search for GDX to find the correct datatable code.
    spy_data = nasdaqdatalink.get_table('ETFG/FUND', ticker='SPY')
    print("Successfully retrieved data for SPY:")
    print(spy_data)

except Exception as e:
    print(f"An error occurred: {e}")

File: nasdaq_data_link.r
# This script demonstrates how to fetch financial data from the Nasdaq Data Link API
# using the httr and jsonlite packages.

# Load required packages
library(httr)
library(jsonlite)
library(dplyr)

# Get API key from environment variable
api_key <- Sys.getenv("NASDAQ_API_KEY")

# Check if API key is set
if (api_key == "") {
    stop("NASDAQ_API_KEY environment variable not set. Please set it to your Nasdaq API key.")
}

# Construct the API URL
ticker <- "SPY"
api_url <- paste0(
    "https://data.nasdaq.com/api/v3/datatables/ETFG/FUND.json?ticker=",
    ticker,
    "&api_key=",
    api_key
)

# Fetch data from the API
cat("Fetching data for", ticker, "ticker...\n")
res <- httr::GET(url = api_url)

# Check if the request was successful
if (httr::status_code(res) == 200) {
    # Parse the JSON response
    json_data <- httr::content(res, as = "text", encoding = "UTF-8")
    data <- jsonlite::fromJSON(json_data)

    # Extract the data table
    spy_data <- data$datatable$data

    # Display the fetched data
    cat("Data fetched successfully:\n")
    print(spy_data)

    # The 'spy_data' object is a data frame that you can now use for analysis.
    # For example, to see the column names:
    cat("\nColumn names:\n")
    print(names(spy_data))
} else {
    # Print an error message if the request failed
    cat("Error fetching data. Status code:", httr::status_code(res), "\n")
    #cat("Response content:\n")
    #print(httr::content(res, as = "text", encoding = "UTF-8"))
}
